Fix for #19479: put display to sleep when idle, prevent firing events…

Heckie75 opened the pull request:
… in DPMS state

## Description
The NotifyXRREvent won't be fired anymore if display is in DPMS state. 

## Motivation and Context
The reason for this change is issue #19479. Displays which are connected via displayport cable send two XRROutputChangeNotifyEvents if they are going to sleep (suspend). First event is of subtype RR_Disconnected. It is followed by an event of subtype RR_Connected. The second event caused that Kodi turns display on again.

I have seen this behaviour on my own display and also other kodi.log files of users who reported this problem which started with Kodi 18.x.

Displays connected via HDMI don't send any events if they are going to sleep.   
 
## How Has This Been Tested?
This fix has been tested on my computer which was connected via displayport. But it was also tested with a connection via HDMI. 

## Screenshots (if appropriate):

## Types of change
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Heckie75 committed:
Fix for #19479: put display to sleep when idle, prevent firing events in DPMS state

lrusak commented on xbmc/windowing/X11/WinEventsX11.cpp:
```diff
@@ -312,6 +312,12 @@ bool CWinEventsX11::MessagePump()
 
+      if (g_application.IsDPMSActive())
+      {
+        CLog::Log(LOGDEBUG,"CWinEventsX11::MessagePump - skip XRRNotifyEvent since DPMS is active");
```
Is this going to end up flooding the log?

Heckie75 replied:
When I tested it in debug mode there was no flooding just a single log entry when the monitor is going into suspend.

lrusak reviewed (approved):
LGTM but I have no knowledge about the x11 window system

Heckie75 commented:
This fix does not work anymore. I can't actually say what the change was. Probably some updates of Linux kernel, Mesa or X11. 

In addition the root cause isn't  Kodi. The problem is mentioned in NVIDIA forum as follows: [DPMS not working with DisplayPort monitor ](https://forums.developer.nvidia.com/t/dpms-not-working-with-displayport-monitor/45144/60)

I found this solution (see [arch linux NVIDIA/Troubleshooting](https://wiki.archlinux.org/title/NVIDIA/Troubleshooting#Driver_415:_HardDPMS)):

_"... This is reported by some users to solve the issues with suspending monitors connected over DisplayPort. It is reported to become the default in a future driver version, but for now, the HardDPMS option can be set in the Device or Screen sections."_

 So I configured my X11 server by adding a config file ```/etc/X11/xorg.conf.d/10-xorg.conf``` including the option  "HardDPMS" "true" within the sections _Monitor_ and _Device_:

```
...
Section "Monitor"
	Identifier   "Monitor0"
	VendorName   "HP"
	ModelName    "U28"
        DisplaySize  621 341
        Option       "DPMS" "true"
        Option       "HardDPMS" "true"
EndSection
...
Section "Device"
	Identifier  "Card0"
	Driver      "intel"
	BusID       "PCI:0:2:0"
        Option      "HardDPMS" "true"
EndSection
...
```
I generated the stub by using ```X -configure``` (see [Where is the X.org config file? How do I configure X there?](https://askubuntu.com/questions/4662/where-is-the-x-org-config-file-how-do-i-configure-x-there))

Although the problem is described to be related to NVIDIA cards it seems to work with the integrated Intel video card _Iris Plus Graphics 655_ as well.

I close this pull request since it does not solve the problem and I found a solution.

Heckie75 commented:
This fix just fixes half of the problem an latest Ubuntu 20.04 and 21.04. I can't actually say what the change was. Probably some updates of Linux kernel, Mesa or X11. In addition to this fix there is also a reconfiguration of X11 required. 

The problem is mentioned in NVIDIA forum as follows: [DPMS not working with DisplayPort monitor ](https://forums.developer.nvidia.com/t/dpms-not-working-with-displayport-monitor/45144/60)

I found this solution (see [arch linux NVIDIA/Troubleshooting](https://wiki.archlinux.org/title/NVIDIA/Troubleshooting#Driver_415:_HardDPMS)):

_"... This is reported by some users to solve the issues with suspending monitors connected over DisplayPort. It is reported to become the default in a future driver version, but for now, the HardDPMS option can be set in the Device or Screen sections."_

 So I configured my X11 server by adding a config file ```/etc/X11/xorg.conf.d/10-xorg.conf``` including the option  "HardDPMS" "true" within the sections _Monitor_ and _Device_:

```
...
Section "Monitor"
	Identifier   "Monitor0"
	VendorName   "HP"
	ModelName    "U28"
        DisplaySize  621 341
        Option       "DPMS" "true"
        Option       "HardDPMS" "true"
EndSection
...
Section "Device"
	Identifier  "Card0"
	Driver      "intel"
	BusID       "PCI:0:2:0"
        Option      "HardDPMS" "true"
EndSection
...
```
I generated the stub by using ```X -configure``` (see [Where is the X.org config file? How do I configure X there?](https://askubuntu.com/questions/4662/where-is-the-x-org-config-file-how-do-i-configure-x-there))

Although the problem is described to be related to NVIDIA cards it seems to work with the integrated Intel video card _Iris Plus Graphics 655_ as well.

However, firstly I thought that the reconfiguration is good enough and there is no fix in Kodi required at all. But then I have seen that Kodi still prevents monitor from going into suspend mode.

Heckie75 commented:
Is there any chance that this PR will be merged for v20 N? Are there any issue that I must solve first?

Heckie75 commented:
Today I have recognized that the problem doesn't exist anymore. Unfortunately I can't say what has solved the problem since I have changed a few things in parallel. I changed from Ubuntu 20.04. to 21.10 with kernel 5.13 and Mesa 21.2.2. The version of Kodi is 19.3 now. And probably the most important change is the display itself. Here I have changed from the HP U28 to the brand new larger version HP U32. 

Because everything is fine for me I have closed the issue #19479 and also this pull request.
